cdist/emulator.py

```diff
@@ -228,40 +256,28 @@ def record_requirements(self):
             for requirement in requirements.split(" "):
                 # Ignore empty fields - probably the only field anyway
                 if len(requirement) == 0: continue
-
-                # Raises an error, if object cannot be created
-                try:
-                    cdist_object = self.cdist_object.object_from_name(requirement)
-                except core.cdist_type.NoSuchTypeError as e:
-                    self.log.error("%s requires object %s, but type %s does not exist. Defined at %s"  % (self.cdist_object.name, requirement, e.name, self.object_source))
-                    raise
-                except core.cdist_object.MissingObjectIdError as e:
-                    self.log.error("%s requires object %s without object id. Defined at %s"  % (self.cdist_object.name, requirement, self.object_source))
-                    raise
-
-                self.log.debug("Recording requirement: %s", requirement)
-
-                # Save the sanitised version, not the user supplied one
-                # (__file//bar => __file/bar)
-                # This ensures pattern matching is done against sanitised list
-                self.cdist_object.requirements.append(cdist_object.name)
-                reqs.add(cdist_object.name)
+                req = self.record_requirement(requirement)
+                reqs.add(req)
         if self._existing_reqs is not None:
             # if object exists then compare existing and new requirements
             self.log.debug("OBJ: {} {}".format(self.cdist_type, self.object_id))
             self.log.debug("EXISTING REQS: {}".format(self._existing_reqs))
             self.log.debug("REQS: {}".format(reqs))
 
             if self._existing_reqs != reqs:
-                errmsg = ("Object {} already exists with conflicting "
-                    "requirements:\n{}: {}\n{}: {}".format(
+                dbgmsg = ("Object {} already exists with different "
+                    "requirements:\n{}: {}\n{}: {}. Merging sets.".format(
                         self.cdist_object.name,
                         " ".join(self.cdist_object.source),
                         self._existing_reqs,
                         self.object_source,
                         reqs))
-                self.log.error(errmsg)
-                raise cdist.Error(errmsg)
+                self.log.debug(dbgmsg)
+                all_reqs = reqs | self._existing_reqs
+                self.log.debug("All requirements: {}".format(all_reqs))
+                for x in all_reqs:
+                    if not x in self.cdist_object.requirements:
+                        self.record_requirement(x)
```

Comment:
@telmich 
I have just (yesterday and today morning) looking at that code and thinking about it.
All this changes about dependency conflict and now this change came from this issue report: https://github.com/ungleich/cdist/issues/441.

I have just contacted Jake to send me his manifests that were causing this. And I have forgotten to add you in CC, so I will include you in email.

Because without above code cdist was already making a union of requirements, and this new code has no effect, but for his manifests it seems that it does not, or there was something wrong with his manifests. Without actual manifests I cannot tell if there is something wrong with cdist code or manifests. And when you look at the code there should be everything ok.
So, according to the present facts, we should change this PR to undo commits for fixing above issue. And when we get Jake's manifests then we can see what was wrong.
What do you think?